CIPs/0009.md

```diff
@@ -0,0 +1,86 @@
+# CIP [0009]: Standalone signing service for Validators
+
+- Date: 2020-05-14
+- Author: @asaj
+- Status: WIP
+
+## Overview
+
+Validator operational excellence is vital to the security and performance of the Celo network.
+
+However, validators currently face significant operational limitations due to key management being bundled into `celo-blockchain`.
+
+In particular, it makes it cumbersome to switch which node is used to participate in consensus.
+
+Switching nodes allows validators to upgrade to newer versions of `celo-blockchain` or to failover to a different node if the primary node fails.
+
+Validators who wish to switch nodes currently have the following options:
+
+1. Bring down node A, bring up node B with key material copied from node A.
+2. Bring up node B with key material copied from node A, bring down node B.
+3. Bring up node B with new key material, authorize the new key, and bring down node A after the next election.
+
+Option `1` results in downtime, option `2` risks double signing, and option `3` can only be done once every 24 hours.
+Furthermore, options `1` and `2` may require unsafe key management practices, which may result in keys being compromised.
+
+Decoupling key management from other validator functionality will allow validators to switch nodes with zero downtime and no risk of double signing.
+
+## Goals
+
+- Zero downtime validator upgrades within an epoch
+- Failover possible with no risk of double signing
+
+## Proposed Solution
+
+The proposed solution is to use `clef`, introduced in [go-ethereum v1.9](https://blog.ethereum.org/2019/07/10/geth-v1-9-0/), as a standalone signer for validators.
+
+`clef` is integrated into `celo-blockchain` as an implementation of the [wallet](https://github.com/celo-org/celo-blockchain/blob/3caca596d1ff23a9faf5339f108c993f38bfa743/accounts/accounts.go#L50) interface in [accounts/external/backend.go](https://github.com/celo-org/celo-blockchain/blob/master/accounts/external/backend.go).
+
+Validators will run a `clef` instance that manages their [validator signer key](https://docs.celo.org/validator-guide/summary/detailed#authorized-validator-signers), and point their node to that instance.
```

Comment:
Clef's security model assumes that the RPC signer link is untrusted and that the user will review and approve every transaction through a UI of some sort. In the validator signer use-case, there will be no user, so presumably Clef would be configured to sign anything it sees and the RPC connection would have to become trusted. Doing so will likely be nontrivial as it goes against the original design assumptions of Clef. 

A quick look at the Clef repo shows that the two options for RPC connection are HTTP and IPC, so either A) support for a secure connection protocol with mutual authentication (e.g. Noise or TLS) will have to added, or B) an external proxy used to proxy the IPC connection over the network. Option B is likely easier, but does add some operational overhead and provides a weaker security boundary.

Once RPC over an encrypted and mutually-authenticated link is supported, switching which validator node is allowed to talk to Clef could be done be removing the public key of the old validator node, and adding the public key of the new one, preferably in an atomic transaction. Upon calling, the old validator should no longer be authorized to make connections and any existing connection should be dropped or at least deauthorized from calling RPCs.

An open question here would be "who is authorized to call initiate a connection handover?" In the Clef model, the UI (`stdin`) Would be the the best candidate, which would essentially require the operator to `ssh` into the machine.

Reply:
If you don't have all the nodes verify all the consensus messages themselves, then you sacrifice some of the security benefits of the (m, n)-threshold signatures. So I think each node should in some sense be a full node, and then the one running clef acts as the aggregator, which puts signatures together from partial signatures.

That said, the nodes may need to exchange some messages besides the partial signatures (which are safe to send in the clear), and for those we probably want mutual authentication (and possibly encryption). Although nodes could sign RPC messages with their ECDSA/ BLS threshold keys, for throughput I agree Noise/TLS is a good option. In your option B do you say a "weaker security boundary" because that option involves some sort of IPC or virtual network activity that might be tampered with by malicious but unprivileged processes on the same machine that wouldn't otherwise be able to do so?

Reply:
I don't mean to suggest that nodes should off-load consensus message processing. Instead, I am suggesting that one way to prevent double signing is to implement some IBFT checks in the Clef sign such that it will refuses to sign two blocks at the same height and round number, to prevent risk of double signing.

To clarify, I am discussing here the communication between the validator nodes and the Clef signer node, which would need to be over an encrypted link. I am not discussing the future case of validator nodes or signer nodes talking between each other to implement a threshold scheme.

On the "weaker security boundary", yes that's exactly what I mean 👍

Reply:
Agreed that clef/node communication will need to be secure.